### PR TITLE
Feature/spainhour/winding number update

### PR DESCRIPTION
**Pull request by jcs15c**
# Summary

This PR is a refactoring + feature aimed towards the current winding number implementations

It makes the following nonfunctional changes:
- Unifies winding number interfaces
  - Consolidates all `winding_number( query, <primal object> )` methods to `winding_number.hpp`
  - Renamed `in_curved_polygon_impl.hpp` to `winding_number_impl.hpp`, with internals unchanged.
- Tightened up `BezierCurve` class implementation to match new conventions in `BezierPatch`, such as more efficient utilization of `axom::Array`

It also adds the following features:
- A robust `primal::Polygon::normal()` method (and corresponding test in `primal_polygon.cpp`)
- `winding_number` overloads for `primal::Triangle<T, 2>`, `primal::Triangle<T, 3>`, `primal::Polygon<T, 3>`, and `primal::Polyhedron<T, 3>` classes along with the requisite tests in `primal_winding_number.cpp` and `primal_solid_angle.cpp`.
-  An `in_polyhedron` method analogous to `in_polygon`, with corresponding test in `primal_polyhedron.cpp`

It also makes the following functional API change:
- Removes the `BezierCurve( T * pts, int ord )` constructor that takes a C-style array of coordinates. As implemented, this is an unintuitive and impractical method of constructing a `BezierCurve` object, and has no analogy for `BezierPatch` objects. Currently, the method is only used in a `primal_bezier_curve.cpp` test, which has been changed.

### CODE REVIEW

**Review by kennyweiss — APPROVED** [review]
This is really nice @jcs15c

Thanks also for adding a comprehensive test suite and working through lots of (literal) corner and edge cases!

Please be sure to add a line or two to the `RELEASE-NOTES` about the new `winding_number` functionality for triangles, polygons and polyhedra.

File: src/axom/primal/geometry/BezierCurve.hpp [hunk 1]
```diff
@@ -234,25 +203,20 @@ class BezierCurve
...
   /// Clears the list of control points, make nonrational
   void clear()
   {
-    const int ord = getOrder();
-    for(int p = 0; p <= ord; ++p)
-    {
-      m_controlPoints[p] = PointType();
-    }
-
+    m_controlPoints.clear();
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/geometry/BezierCurve.hpp:219]: Good catch, there actually isn't a test for `BezierCurve::clear()` at all. I've added one now.

File: src/axom/primal/geometry/Polygon.hpp [hunk 2]
```diff
@@ -79,6 +80,30 @@ class Polygon
...
+
+    VectorType normal, v0(m_vertices[0], m_vertices[1]), v1;
+
+    // Iterate over each pair of vertices
+    for(int i = 2; i < nverts; ++i)
+    {
+      v1 = VectorType(m_vertices[0], m_vertices[i]);
+      normal += VectorType::cross_product(v0, v1);
+      v0 = v1;
+    }
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/geometry/Polygon.hpp:102]: This algorithm comes directly from triangulating the polygon and computing the normal vector for each one, so it's already taking those areas into consideration implicitly. Small triangles have small spanning vectors, which leads to a small magnitude cross product.

[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/geometry/Polygon.hpp:102]: Actually, I naively copied this algorithm from the Graphics Gem about polyhedron in-out, but this isn't actually the best way to compute the normal vector. As I mentioned in the other comment, this algorithm is equivalent to computing normal vectors in each triangulation, but if you expand out the algebra of those cross products and cancel them out properly, you get an algorithm that iterates over consecutive pairs of edges instead of vectors from $V_0$ to $V_i$ and $V_{i+1}$. I'll update the algorithm in a bit. In the end, it'll make the algorithm independent of that choice of triangulation.

[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/primal/geometry/Polygon.hpp:102]: > it's already taking those areas into consideration implicitly

Of course -- good call!

File: src/axom/primal/operators/winding_number.hpp [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+      if(axom::utilities::isNearlyEqual(P[j][0], R[0], EPS))
+        return !useStrictInclusion;  // On vertex
+      else if(P[i][1] == R[1] && ((P[j][0] > R[0]) == (P[i][0] < R[0])))
+        return !useStrictInclusion;  // On horizontal edge
+    }
+
+    // Check if edge crosses horizontal line
+    if((P[i][1] < R[1]) != (P[j][1] < R[1]))
+    {
+      double det;
```
Comment (gunney1) [inline src/axom/primal/operators/winding_number.hpp:122]: It would be cleaner and less confusing to read if det was declared in the 2 scopes where it's used.

File: src/axom/primal/geometry/Polygon.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/geometry/Polygon.hpp]: Updated algorithm is in!

File: src/axom/primal/operators/detail/winding_number_impl.hpp [hunk 4]
```diff
@@ -165,39 +167,45 @@ double convex_endpoint_winding_number(const Point<T, 2>& q,
...
 
-  // Use linearity as base case for recursion
+  // If q is outside a convex shape that contains the entire curve, the winding
+  //   number for the shape connected at the endpoints with straight lines is zero.
+  //   We then subtract the contribution of this line segment.
+
+  // Simplest convex shape containing c is its bounding box
+  BoundingBox<T, 2> bBox(c.boundingBox());
+  if(!bBox.contains(q))
+    return 0.0 - linear_winding_number(q, c[ord], c[0], edge_tol);
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/detail/winding_number_impl.hpp:194]: This was discussed irl, but I'll post the general idea here as well.

The exterior of any closed shape has a winding number of zero, even if the curve is non-convex or simple. This means that the simplest and most general test we can do is check a bounding box for the curve. If we're outside that bounding box, we can always close the curve with a line and use that value in our calculation, regardless of what the curve is doing.

File: src/axom/primal/operators/winding_number.hpp [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+ * \brief Computes the solid angle winding number for a 3D planar polygon
+ *
+ * \param [in] query The query point to test
+ * \param [in] poly The Polygon object
+ * \param [in] isOnFace Return variable to show if the point is on the polygon
+ * \param [in] edge_tol The physical distance level at which objects are 
+ *                      considered indistinguishable
+ * \param [in] EPS Miscellaneous numerical tolerance level for nonphysical distances
+ * 
+ * \pre Assumes the polygon is truly planar
```
[the remark this answers is not in the document]
Comment (gunney1) [reply src/axom/primal/operators/winding_number.hpp:293]: Looks like it would use an average normal, which seems reasonable if polygon is not too non-planar.

File: src/axom/primal/operators/winding_number.hpp [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+int winding_number(const Point<T, 2>& q,
+                   const Triangle<T, 2>& tri,
+                   bool useStrictInclusion = false,
+                   double edge_tol = 1e-8)
+{
+  return winding_number(
+    q,
+    Polygon<T, 2>(axom::Array<Point<T, 2>>({tri[0], tri[1], tri[2]})),
+    useStrictInclusion,
+    edge_tol);
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:71]: This is the best optimization I can think of. While it's possible to construct the winding number for the triangle as a sum of linear winding numbers for each edge, it's easiest to just check containment + orientation as a polygon. The `winding_number( Polygon )` overload is, I believe anyway, to be the most efficient way of doing this, as it can (usually) solve the problem just with floating point comparisons. 

There actually is a `Triangle::checkInTriangle( Point )` method that accomplishes the same thing, but I believe that method is less efficient, since it converts the query to barycentric coordinates first, which it computes directly as ratios of areas. Even in the best case, that involves more floating point multiplications.

File: src/axom/primal/operators/winding_number.hpp [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+ * \param [in] poly The Polygon object
+ * \param [in] isOnFace Return variable to show if the point is on the polygon
+ * \param [in] edge_tol The physical distance level at which objects are 
+ *                      considered indistinguishable
+ * \param [in] EPS Miscellaneous numerical tolerance level for nonphysical distances
+ * 
+ * \pre Assumes the polygon is truly planar
+ * Triangulates the polygon and computes the triangular solid angle for each part
+ * 
+ * \return float the generalized winding number.
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:296]: Fixed here and elsewhere

File: src/axom/primal/operators/winding_number.hpp [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+ * \pre Expects the polyhedron to be convex
+ * 
+ * Computes the faces of the polygon and computes the winding number for each.
+ *
+ * \return int The integer winding number.
+ */
+template <typename T>
+int winding_number(const Point<T, 3>& query,
+                   const Polyhedron<T, 3>& poly,
+                   bool useStrictInclusion = false,
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:353]: This has the opposite meaning though, does it not? If `useStrictInclusion == true` then vertices are considered exterior, but if `includeBoundary == true`, then vertices are considered interior. That's not to say that the convention shouldn't be made consistent, just that renaming would involve a functional change to the behavior.

File: src/axom/primal/operators/winding_number.hpp [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+                   double edge_tol = 1e-8,
+                   double EPS = 1e-8)
+{
+  SLIC_ASSERT(poly.hasNeighbors());
+  const int num_verts = poly.numVertices();
+
+  int* faces = new int[num_verts * num_verts];
+  int* face_size = new int[2 * num_verts];
+  int* face_offset = new int[2 * num_verts];
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:362]: This is only the way that it is so that the data is compatible with the later call to `Polyhedron::getFaces`. Is there a way to access an underlying integer array from an `axom::Array<int>`?

I'm also not opposed to working towards changing/overloading `Polyhedron::getFaces` to take in `axom::Array` arguments, if that's something that would be useful.

File: src/axom/primal/operators/winding_number.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/primal/operators/winding_number.hpp:353]: oops -- I interpreted `useStrictInclusion` to mean that the boundary is interior.
Looks like I should have paid more attention :).

In that case, do you think `includeBoundary` might more obviously convey its intention than `useStrictInclusion` ?

Anyone else have an opinion?

File: src/axom/primal/operators/winding_number.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/primal/operators/winding_number.hpp:362]: I think `Array::data()` should do the trick

https://github.com/LLNL/axom/blob/31ae2e4787a602d9a7b44eea76c3130a30c8ce3d/src/axom/core/Array.hpp#L339-L340

File: src/axom/primal/operators/winding_number.hpp [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+ * \param [in] EPS Miscellaneous numerical tolerance level for nonphysical distances
+ * 
+ * \pre Expects the polyhedron to be convex
+ * 
+ * Computes the faces of the polygon and computes the winding number for each.
+ *
+ * \return int The integer winding number.
+ */
+template <typename T>
+int winding_number(const Point<T, 3>& query,
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:351]: That's correct, I'm assuming the polyhedron is closed (the documentation even assumes that the polyhedron is convex, which doesn't need to be true for this algorithm to work). I'll add some lines in the documentation that address this, but the other `winding_number` functions that operate on closed shapes (such as 2D triangles and polygons) return integers as well.

File: src/axom/primal/operators/winding_number.hpp [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+ * \brief Computes the solid angle winding number for a 3D convex polyhedron
+ *
+ * \param [in] query The query point to test
+ * \param [in] poly The Polyhedron object
+ * \param [in] useStrictInclusion If true, points on the boundary are considered exterior.
+ * \param [in] edge_tol The physical distance level at which objects are 
+ *                      considered indistinguishable
+ * \param [in] EPS Miscellaneous numerical tolerance level for nonphysical distances
+ * 
+ * \pre Expects the polyhedron to be convex
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:344]: Convexity is something that's assumed for every `primal::Polyhedron` object (but isn't checked for in `Polyhedron::isValid`, and is in general a nontrivial computation), but the algorithm here would work even if that wasn't the case. I have the note here to clarify this, mostly to ensure that the requisite `Polyhedron::getFaces` method is guaranteed to work.

File: src/axom/primal/operators/winding_number.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:362]: Ah, I was searching in `Array.hpp` for something with `get` in the name. I'll make the change, thank you!

File: src/axom/primal/operators/winding_number.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (gunney1) [reply src/axom/primal/operators/winding_number.hpp:353]: I agree with @kennyweiss  that `includeBoundary` more obviously convey its intention than `useStrictInclusion`.

File: src/axom/primal/geometry/BezierCurve.hpp [hunk 12]
```diff
@@ -139,10 +107,7 @@ class BezierCurve
     const int sz = utilities::max(0, ord + 1);
     m_controlPoints.resize(sz);
 
-    for(int p = 0; p <= ord; ++p)
-    {
-      m_controlPoints[p] = pts[p];
-    }
+    for(int p = 0; p <= ord; ++p) m_controlPoints[p] = pts[p];
```
Comment (rhornung67) [inline src/axom/primal/geometry/BezierCurve.hpp:110]:
```suggestion
    for(int p = 0; p <= ord; ++p) {
      m_controlPoints[p] = pts[p];
    }
```

I believe our style guide requires this.

File: src/axom/primal/geometry/BezierCurve.hpp [hunk 13]
```diff
@@ -166,7 +131,7 @@ class BezierCurve
     for(int p = 0; p <= ord; ++p) m_controlPoints[p] = pts[p];
 
     if(weights == nullptr)
-      m_weights.resize(0);
+      makeNonrational();
```
Comment (rhornung67) [inline src/axom/primal/geometry/BezierCurve.hpp:134]:
```suggestion
      makeNonrational();
```

Comment (rhornung67) [inline src/axom/primal/geometry/BezierCurve.hpp:134]:
```suggestion
      { 
         makeNonrational();
      }
```

File: src/axom/primal/geometry/BezierPatch.hpp [hunk 14]
```diff
@@ -453,7 +453,7 @@ class BezierPatch
 
     if(isRational())
     {
-      NumsMat new_weights(ord_v + 1, ord_u + 1);
+      WeightsMat new_weights(ord_v + 1, ord_u + 1);
       for(int p = 0; p <= ord_u; ++p)
         for(int q = 0; q <= ord_v; ++q) new_weights(q, p) = m_weights(p, q);
```
Comment (rhornung67) [inline src/axom/primal/geometry/BezierPatch.hpp:458]:
```suggestion
        for(int q = 0; q <= ord_v; ++q) 
        { 
          new_weights(q, p) = m_weights(p, q);
        }
```

File: src/axom/primal/tests/primal_polygon.cpp [hunk 15]
```diff
@@ -369,6 +369,88 @@ TEST(primal_polygon, area_2d_3d)
...
+    Vector3D v1 = Vector3D({0.0, 1.0, 2.0}).unitVector();
+    Vector3D v2 = Vector3D({2.0, -1.0, 0.5}).unitVector();
+    Vector3D exp_normal = Vector3D::cross_product(v1, v2).unitVector();
+
+    Polygon3D poly(5);
+
+    double angles[5] = {0.0, 0.5, 1.2, 3.0, 5.0};
+
+    for(int i = 0; i < 5; ++i)
+      poly.addVertex(Point3D {cos(angles[i]) * v1[0] + sin(angles[i]) * v2[0],
```
Comment (rhornung67) [inline src/axom/primal/tests/primal_polygon.cpp:404]:
```suggestion
      {
        poly.addVertex(Point3D {cos(angles[i]) * v1[0] + sin(angles[i]) * v2[0],
```

and close it off with a vertically aligned '}'

File: src/axom/primal/tests/primal_polygon.cpp [hunk 16]
```diff
@@ -369,6 +369,88 @@ TEST(primal_polygon, area_2d_3d)
...
+    // Add a midpoint between angles 0 and 1
+    poly.addVertex(Point3D::midpoint(
+      poly[0],
+      Point3D {cos(angles[1]) * v1[0] + sin(angles[1]) * v2[0],
+               cos(angles[1]) * v1[1] + sin(angles[1]) * v2[1],
+               cos(angles[1]) * v1[2] + sin(angles[1]) * v2[2]}));
+
+    // Add the rest of the vertices
+    for(int i = 1; i < 8; ++i)
+      poly.addVertex(Point3D {cos(angles[i]) * v1[0] + sin(angles[i]) * v2[0],
```
Comment (rhornung67) [inline src/axom/primal/tests/primal_polygon.cpp:436]:
```suggestion
      {
         poly.addVertex(Point3D {cos(angles[i]) * v1[0] + sin(angles[i]) * v2[0],
```

and close with '}'

File: src/axom/primal/tests/primal_polygon.cpp [hunk 17]
```diff
@@ -369,6 +369,88 @@ TEST(primal_polygon, area_2d_3d)
...
+    poly.addVertex(Point3D::midpoint(poly[0], poly[8]));
+
+    // Verify degeneracies
+    EXPECT_NEAR(0.0, Triangle3D(poly[0], poly[1], poly[2]).area(), 1e-10);
+    EXPECT_NEAR(0.0, Triangle3D(poly[2], poly[3], poly[4]).area(), 1e-10);
+
+    Vector3D obs_normal = poly.normal().unitVector();
+
+    EXPECT_EQ(10, poly.numVertices());
+    for(int i = 0; i < 3; ++i) EXPECT_NEAR(exp_normal[i], obs_normal[i], 1e-10);
```
Comment (rhornung67) [inline src/axom/primal/tests/primal_polygon.cpp:450]:
```suggestion
    for(int i = 0; i < 3; ++i) 
    {
      EXPECT_NEAR(exp_normal[i], obs_normal[i], 1e-10);
    }
```

File: src/axom/primal/tests/primal_polyhedron.cpp [hunk 18]
```diff
@@ -96,6 +101,45 @@ TEST(primal_polyhedron, polyhedron_unit_cube)
...
+  bool edge_tol = 0.0;
+  bool EPS = 0.0;
+
+  for(double x = -0.5; x <= 1.5; x += 0.1)
+    for(double y = -0.5; y <= 1.5; y += 0.1)
+      for(double z = -0.5; z <= 1.5; z += 0.1)
+      {
+        if((x >= 0.0 && x <= 1.0) && (y >= 0.0 && y <= 1.0) &&
+           (z >= 0.0 && z <= 1.0))
+          EXPECT_TRUE(in_polyhedron(PointType({x, y, z}),
```
Comment (rhornung67) [inline src/axom/primal/tests/primal_polyhedron.cpp:119]:
```suggestion
          { 
            EXPECT_TRUE(in_polyhedron(PointType({x, y, z}),
```

File: src/axom/primal/tests/primal_polyhedron.cpp [hunk 19]
```diff
@@ -127,6 +171,32 @@ TEST(primal_polyhedron, polyhedron_tetrahedron)
...
+    bool useNonzeroRule = true;
+
+    // Use zero numerical tolerances
+    bool edge_tol = 0.0;
+    bool EPS = 0.0;
+
+    for(double z = -1.5; z < 1.5; z += 0.1)
+    {
+      if((z >= -1.0) && (z <= 1.0))
+        EXPECT_TRUE(in_polyhedron(PointType({0.0, 0.0, z}),
```
Comment (rhornung67) [inline src/axom/primal/tests/primal_polyhedron.cpp:186]: enclose each part of if-else clause inside braces

**Review by gunney1 — APPROVED** [review]
Thank you @jcs15c

File: src/axom/primal/operators/winding_number.hpp [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+
+// C++ includes
+#include <cmath>
+
+namespace axom
+{
+namespace primal
+{
+/*
+ * \brief Compute the winding number with respect to a line segment
```
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:36]: There could conceivably be a function `winding_number( Segment<T, 3> )`, but it would always return 0. I'll definitely clarify things in the doxystring, but I also like the idea of separating the 2D and 3D functions with doxygen sections. Is it still possible to have the individual doxygen headers for each function inside the larger group?

File: src/axom/primal/geometry/BezierCurve.hpp (hunk 12 above)
Comment (kennyweiss) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: I'm not sure that if it does, although I agree that we probably want it to.

If we want to enforce curly braces for one-lines, we should probably update our clang-format rules to do this automatically in a separate PR.

Comment (kennyweiss) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: Never mind -- it's right at the top of our code formatting -- 
> 9.1 Curly braces should be used in all conditionals, loops, etc. even when the content inside the braces is a “one-liner”.
https://axom.readthedocs.io/en/develop/docs/sphinx/coding_guide/sec09_format.html

We should modify our clang-format to enforce this.

Comment (rhornung67) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: That would be good, if it's supported.

BTW, thanks for confirming. I thought I may be hallucinating.

File: src/axom/primal/operators/winding_number.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/primal/operators/winding_number.hpp:36]: Yes -- the individual doxygen comments remain, but it adds an extra level of grouping.

Here's an example from `sidre::Group`: 

image
https://axom.readthedocs.io/en/develop/doxygen/html/classaxom_1_1sidre_1_1Group.html

File: src/axom/primal/operators/winding_number.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:353]: The more tests I write that use the parameter, the more I agree with it too. Thanks for the suggestion!

File: src/axom/primal/geometry/BezierCurve.hpp (hunk 12 above)
Comment (kennyweiss) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: Looks like it was added in clang@15 and we're on clang@10 for our `clang-format`, but we could use clang-tidy in the meantime.

See: https://stackoverflow.com/questions/26111162/can-clang-format-add-braces-to-single-line-if-statements-etc

Comment (rhornung67) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: Unfortunately, the latest clang on any LC system is clang@14.0.6 (TOSS4).

File: src/axom/primal/operators/winding_number.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:360]: Thinking about this further, our `in_curved_polygon` method from last summer has identical behavior to `in_polygon` and `in_polyhedron`, but `winding_number( CurvedPolygon )` returns a `double` while `winding_number( Polygon/Polyhedron)` return an `int`. Is that an inconsistency, or unintuitive even? 

Right now the first thing `in_curved_polygon` does is round the return value of `winding_number( CurvedPolygon )`, but I'm half tempted to make `winding_number( CurvedPolygon )` assume/check that the input is closed (up to a certain tolerance) and return an `int` like the other methods. A `primal::CurvedPolygon` object is just a collection of Bezier curve edges, but if it isn't closed, then it isn't really a polygon, right?

File: src/axom/primal/geometry/BezierCurve.hpp (hunk 12 above)
Comment (jcs15c) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: I'll admit, I did not know that was part of the style guide. I'll go through and make all the changes I see...

Comment (rhornung67) [reply src/axom/primal/geometry/BezierCurve.hpp:110]: No worries. While the curly braces are not necessary, they do make the code easier to understand, arguably. In the future, we will add this to our code formatter.

File: src/axom/primal/operators/winding_number.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jcs15c) [reply src/axom/primal/operators/winding_number.hpp:333]: This is correct, and is where a good amount of the "robustness" of the algorithm comes from.

**Review by rhornung67 — APPROVED** [review]
Thank you @jcs15c

File: src/axom/primal/geometry/BezierCurve.hpp (hunk 12 above)
Comment (kennyweiss) [reply src/axom/primal/geometry/BezierCurve.hpp]: I created an issue for this: https://github.com/LLNL/axom/issues/1117

**Comment by jcs15c** [comment]
Thank you everyone for the comments and advice!
